File: core/src/main/java/at/stefangeyer/challonge/service/implementation/SimpleParticipantService.java
package at.stefangeyer.challonge.service.implementation;

import at.stefangeyer.challonge.async.Callback;
import at.stefangeyer.challonge.exception.DataAccessException;
import at.stefangeyer.challonge.model.Participant;
import at.stefangeyer.challonge.model.Tournament;
import at.stefangeyer.challonge.model.query.ParticipantQuery;
import at.stefangeyer.challonge.model.query.wrapper.ParticipantQueryListWrapper;
import at.stefangeyer.challonge.model.query.wrapper.ParticipantQueryWrapper;
import at.stefangeyer.challonge.model.wrapper.ParticipantWrapper;
import at.stefangeyer.challonge.rest.ParticipantRestClient;
import at.stefangeyer.challonge.service.ParticipantService;

import java.util.List;
import java.util.stream.Collectors;

/**
 * Participant Service Implementation
 *
 * @author Stefan Geyer
 * @version 2018-06-30
 */
public class SimpleParticipantService implements ParticipantService {

    private ParticipantRestClient restClient;

    public SimpleParticipantService(ParticipantRestClient restClient) {
        this.restClient = restClient;
    }

    @Override
    public List<Participant> getParticipants(Tournament tournament) throws DataAccessException {
        return this.restClient.getParticipants(String.valueOf(tournament.getId())).stream()
                .map(ParticipantWrapper::getParticipant).collect(Collectors.toList());
    }

    @Override
    public void getParticipants(Tournament tournament, Callback<List<Participant>> onSuccess, Callback<DataAccessException> onFailure) {
        this.restClient.getParticipants(String.valueOf(tournament.getId()), list -> onSuccess.accept(list.stream()
                .map(ParticipantWrapper::getParticipant).collect(Collectors.toList())), onFailure);
    }

    @Override
    public Participant getParticipant(Tournament tournament, long participantId, boolean includeMatches) throws DataAccessException {
        return this.restClient.getParticipant(String.valueOf(tournament.getId()), participantId, includeMatches).getParticipant();
    }

    @Override
    public void getParticipant(Tournament tournament, long participantId, boolean includeMatches,
                               Callback<Participant> onSuccess, Callback<DataAccessException> onFailure) {
        this.restClient.getParticipant(String.valueOf(tournament.getId()), participantId, includeMatches,
                pw -> onSuccess.accept(pw.getParticipant()), onFailure);
    }

    @Override
    public Participant addParticipant(Tournament tournament, ParticipantQuery data) throws DataAccessException {
        validateParticipantQuery(data);
        return this.restClient.addParticipant(String.valueOf(tournament.getId()), new ParticipantQueryWrapper(data)).getParticipant();
    }

    @Override
    public void addParticipant(Tournament tournament, ParticipantQuery data, Callback<Participant> onSuccess,
                               Callback<DataAccessException> onFailure) {
        validateParticipantQuery(data);
        this.restClient.addParticipant(String.valueOf(tournament.getId()), new ParticipantQueryWrapper(data),
                pw -> onSuccess.accept(pw.getParticipant()), onFailure);
    }

    @Override
    public List<Participant> bulkAddParticipants(Tournament tournament, List<ParticipantQuery> data) throws DataAccessException {
        for (ParticipantQuery query : data) {
            validateParticipantQuery(query);
        }

        return this.restClient.bulkAddParticipants(String.valueOf(tournament.getId()),
                new ParticipantQueryListWrapper(data)).stream().map(ParticipantWrapper::getParticipant)
                .collect(Collectors.toList());
    }

    @Override
    public void bulkAddParticipants(Tournament tournament, List<ParticipantQuery> data,
                                    Callback<List<Participant>> onSuccess, Callback<DataAccessException> onFailure) {
        for (ParticipantQuery query : data) {
            validateParticipantQuery(query);
        }

        this.restClient.bulkAddParticipants(String.valueOf(tournament.getId()), new ParticipantQueryListWrapper(data),
                list -> onSuccess.accept(list.stream().map(ParticipantWrapper::getParticipant).collect(Collectors.toList())), onFailure);
    }

    @Override
    public Participant updateParticipant(Participant participant, ParticipantQuery data) throws DataAccessException {
        validateParticipantQuery(data);

        return this.restClient.updateParticipant(String.valueOf(participant.getTournamentId()), participant.getId(),
                new ParticipantQueryWrapper(data)).getParticipant();
    }

    @Override
    public void updateParticipant(Participant participant, ParticipantQuery data, Callback<Participant> onSuccess,
                                  Callback<DataAccessException> onFailure) {
        validateParticipantQuery(data);

        this.restClient.updateParticipant(String.valueOf(participant.getTournamentId()), participant.getId(),
                new ParticipantQueryWrapper(data), pw -> onSuccess.accept(pw.getParticipant()), onFailure);
    }

    @Override
    public Participant checkInParticipant(Participant participant) throws DataAccessException {
        return this.restClient.checkInParticipant(String.valueOf(participant.getTournamentId()), participant.getId()).getParticipant();
    }

    @Override
    public void checkInParticipant(Participant participant, Callback<Participant> onSuccess, Callback<DataAccessException> onFailure) {
        this.restClient.checkInParticipant(String.valueOf(participant.getTournamentId()), participant.getId(),
                pw -> onSuccess.accept(pw.getParticipant()), onFailure);
    }

    @Override
    public Participant undoCheckInParticipant(Participant participant) throws DataAccessException {
        return this.restClient.undoCheckInParticipant(String.valueOf(participant.getTournamentId()), participant.getId()).getParticipant();
    }

    @Override
    public void undoCheckInParticipant(Participant participant, Callback<Participant> onSuccess, Callback<DataAccessException> onFailure) {
        this.restClient.undoCheckInParticipant(String.valueOf(participant.getTournamentId()), participant.getId(),
                pw -> onSuccess.accept(pw.getParticipant()), onFailure);
    }

    @Override
    public Participant deleteParticipant(Participant participant) throws DataAccessException {
        return this.restClient.deleteParticipant(String.valueOf(participant.getTournamentId()), participant.getId()).getParticipant();
    }

    @Override
    public void deleteParticipant(Participant participant, Callback<Participant> onSuccess, Callback<DataAccessException> onFailure) {
        this.restClient.deleteParticipant(String.valueOf(participant.getTournamentId()), participant.getId(),
                pw -> onSuccess.accept(pw.getParticipant()), onFailure);
    }

    @Override
    public List<Participant> randomizeParticipants(Tournament tournament) throws DataAccessException {
        return this.restClient.randomizeParticipants(String.valueOf(tournament.getId())).stream()
                .map(ParticipantWrapper::getParticipant).collect(Collectors.toList());
    }

    @Override
    public void randomizeParticipants(Tournament tournament, Callback<List<Participant>> onSuccess, Callback<DataAccessException> onFailure) {
        this.restClient.randomizeParticipants(String.valueOf(tournament.getId()),
                list -> onSuccess.accept(list.stream().map(ParticipantWrapper::getParticipant).collect(Collectors.toList())), onFailure);
    }

    private void validateParticipantQuery(ParticipantQuery data) {
        if (data.getName() == null && data.getEmail() == null && data.getChallongeUsername() == null &&
                data.getSeed() == null && data.getMisc() == null && data.getInviteNameOrEmail() == null) {
            throw new IllegalArgumentException("All data parameters are null. Provide at least one");
        }
    }
}

File: core/src/main/kotlin/at/stefangeyer/challonge/service/implementation/SimpleParticipantService.kt
package at.stefangeyer.challonge.service.implementation

import at.stefangeyer.challonge.async.Callback
import at.stefangeyer.challonge.exception.DataAccessException
import at.stefangeyer.challonge.model.Participant
import at.stefangeyer.challonge.model.Tournament
import at.stefangeyer.challonge.model.query.ParticipantQuery
import at.stefangeyer.challonge.model.query.wrapper.ParticipantQueryListWrapper
import at.stefangeyer.challonge.model.query.wrapper.ParticipantQueryWrapper
import at.stefangeyer.challonge.rest.ParticipantRestClient
import at.stefangeyer.challonge.service.ParticipantService

/**
 * Participant Service Implementation
 *
 * @author Stefan Geyer
 * @version 2018-06-30
 */
class SimpleParticipantService(private val restClient: ParticipantRestClient) : ParticipantService {

    override fun getParticipants(tournament: Tournament): List<Participant> =
            this.restClient.getParticipants(tournament.id.toString()).map { pw -> pw.participant }

    override fun getParticipants(tournament: Tournament,
                                 onSuccess: Callback<List<Participant>>, onFailure: Callback<DataAccessException>) {
        this.restClient.getParticipants(tournament.id.toString(),
                Callback { list -> onSuccess.accept(list.map { pw -> pw.participant }) }, onFailure)
    }

    override fun getParticipant(tournament: Tournament, participantId: Long, includeMatches: Boolean): Participant =
            this.restClient.getParticipant(tournament.id.toString(), participantId, includeMatches).participant

    override fun getParticipant(tournament: Tournament, participantId: Long, includeMatches: Boolean,
                                onSuccess: Callback<Participant>, onFailure: Callback<DataAccessException>) {
        this.restClient.getParticipant(tournament.id.toString(), participantId, includeMatches,
                Callback { pw -> onSuccess.accept(pw.participant) }, onFailure)
    }

    override fun addParticipant(tournament: Tournament, data: ParticipantQuery): Participant {
        validateParticipantQuery(data)
        return this.restClient.addParticipant(tournament.id.toString(), ParticipantQueryWrapper(data)).participant
    }

    override fun addParticipant(tournament: Tournament, data: ParticipantQuery,
                                onSuccess: Callback<Participant>, onFailure: Callback<DataAccessException>) {
        validateParticipantQuery(data)
        this.restClient.addParticipant(tournament.id.toString(), ParticipantQueryWrapper(data),
                Callback { pw -> onSuccess.accept(pw.participant) }, onFailure)
    }

    override fun bulkAddParticipants(tournament: Tournament, data: List<ParticipantQuery>): List<Participant> {
        for (query in data) {
            validateParticipantQuery(query)
        }
        return this.restClient.bulkAddParticipants(tournament.id.toString(),
                ParticipantQueryListWrapper(data)).map { pq -> pq.participant }
    }

    override fun bulkAddParticipants(tournament: Tournament, data: List<ParticipantQuery>,
                                     onSuccess: Callback<List<Participant>>, onFailure: Callback<DataAccessException>) {
        for (query in data) {
            validateParticipantQuery(query)
        }
        this.restClient.bulkAddParticipants(
                tournament.id.toString(), ParticipantQueryListWrapper(data),
                Callback { list -> onSuccess.accept(list.map { pw -> pw.participant }) }, onFailure)
    }

    override fun updateParticipant(participant: Participant, data: ParticipantQuery): Participant {
        validateParticipantQuery(data)
        return this.restClient.updateParticipant(participant.tournamentId.toString(), participant.id,
                ParticipantQueryWrapper(data)).participant
    }

    override fun updateParticipant(participant: Participant, data: ParticipantQuery,
                                   onSuccess: Callback<Participant>, onFailure: Callback<DataAccessException>) {
        validateParticipantQuery(data)
        return this.restClient.updateParticipant(participant.tournamentId.toString(), participant.id,
                ParticipantQueryWrapper(data), Callback { pw -> onSuccess.accept(pw.participant) }, onFailure)
    }

    override fun checkInParticipant(participant: Participant): Participant =
            this.restClient.checkInParticipant(participant.tournamentId.toString(), participant.id).participant

    override fun checkInParticipant(participant: Participant,
                                    onSuccess: Callback<Participant>, onFailure: Callback<DataAccessException>) {
        this.restClient.checkInParticipant(participant.tournamentId.toString(), participant.id,
                Callback { pw -> onSuccess.accept(pw.participant) }, onFailure)
    }

    override fun undoCheckInParticipant(participant: Participant): Participant =
            this.restClient.undoCheckInParticipant(participant.tournamentId.toString(), participant.id).participant

    override fun undoCheckInParticipant(participant: Participant,
                                        onSuccess: Callback<Participant>, onFailure: Callback<DataAccessException>) {
        this.restClient.undoCheckInParticipant(participant.tournamentId.toString(), participant.id,
                Callback { pw -> onSuccess.accept(pw.participant) }, onFailure)
    }

    override fun deleteParticipant(participant: Participant): Participant =
            this.restClient.deleteParticipant(participant.tournamentId.toString(), participant.id).participant

    override fun deleteParticipant(participant: Participant,
                                   onSuccess: Callback<Participant>, onFailure: Callback<DataAccessException>) {
        this.restClient.deleteParticipant(participant.tournamentId.toString(), participant.id,
                Callback { pw -> onSuccess.accept(pw.participant) }, onFailure)
    }

    override fun randomizeParticipants(tournament: Tournament): List<Participant> =
            this.restClient.randomizeParticipants(tournament.id.toString()).map { pw -> pw.participant }

    override fun randomizeParticipants(tournament: Tournament,
                                       onSuccess: Callback<List<Participant>>, onFailure: Callback<DataAccessException>) {
        this.restClient.randomizeParticipants(tournament.id.toString(),
                Callback { list -> onSuccess.accept(list.map { pw -> pw.participant }) }, onFailure)
    }

    private fun validateParticipantQuery(data: ParticipantQuery) {
        if (data.name == null && data.email == null && data.challongeUsername == null &&
                data.seed == null && data.misc == null && data.inviteNameOrEmail == null) {
            throw IllegalArgumentException("All data parameters are null. Provide at least one")
        }
    }
}